File: Book/again/09-2. 전보.js
const n = 3;
const c = 1;
const info = [
  [1, 2, 4],
  [1, 3, 2],
];

function solution(n, c, info) {
  const INF = 1e9;
  // 초기 배열 설정
  const graph = Array.from(Array(n + 1), () => Array(0));
  // 간선 정보 등록
  info.forEach((information) => {
    const [a, b, cost] = information;
    graph[a].push([cost, b]);
  });
  // 거리 정보 등록할 배열
  const distance = Array(n + 1).fill(INF);

  function dijkstra(start) {
    // 노드를 담을 배열 queue
    const queue = [];
    queue.push([0, start]);
    // 시작점 0으로 표시
    distance[start] = 0;
    // 연결된 부분 확인 queue가 비워질 때까지 계속 한다.
    while (queue.length > 0) {
      // 먼저 queue를 cost에 따라 정렬해주어야 한다.
      queue.sort((a, b) => a[0] - b[0]);
      // queue에서 뺀 노드
      const [currentCost, currentNode] = queue.shift();
      // 해당 노드에 연결되어있는 노드들을 체크한다.
      for (let i = 0; i < graph[currentNode].length; i++) {
        const [tempCost, tempNext] = graph[currentNode][i];
        distance[tempNext] = Math.min(
          distance[tempNext],
          tempCost + currentCost
        );
        queue.push([distance[tempNext], tempNext]);
      }
    }
  }
  dijkstra(c);
  // 결과를 담을 변수
  let count = 0;
  let time = 0;
  // distance를 순회하면서 INF와 0이 아닌 값들을 세서 count에 넣고, 그 중 가장 큰 값을 time에 넣는다.
  for (let i = 1; i < n + 1; i++) {
    if (distance[i] > 0 && distance[i] < INF) {
      count++;
      if (time < distance[i]) time = distance[i];
    }
  }
  return [count, time];
}

console.log(solution(n, c, info));

File: Book/again/09-2. 전보.py
import pprint as pp
import heapq

n, m, c = map(int, input().split())  # 노드, 간선, 도시
INF = int(1e9)
graph = [[] for _ in range(n+1)]    # 정보를 담을 배열 graph
distance = [INF] * (n+1)    # 거리를 담을 distance를 INF 값으로 초기화


for _ in range(m):  # 간선 정보 입력받기
    x, y, z = map(int, input().split())
    graph[x].append((y, z))  # 출발점x에 도착점y와 cost를 튜플 형태로 추가


def dijkstra(start):
    queue = []
    heapq.heappush(queue, (0, start))   # 우선순위큐를 만들고 (cost, 시작점) 추가
    distance[start] = 0  # 시작점 cost = 0

    while queue:
        dist, now = heapq.heappop(queue)    # 우선순위큐에서 하나 빼줌
        if distance[now] < dist:    # 현재 distance가 저장되어있는 distance보다 더 크면 skip
            continue
        for i in graph[now]:    # 현재 연결되어 있는 노드 순회
            cost = dist + i[1]  # 거리가 1이므로 dist + 1

            # 현재 distance(cost)가 저장되어있는 distance보다 작으면
            if cost < distance[i[0]]:
                distance[i[0]] = cost   # distance 업데이트
                heapq.heappush(queue, (cost, i[0]))  # 큐에 추가


dijkstra(c)
pp.pprint(distance)

count = 0   # 개수
max_distance = 0    # 가장 먼거리

for dist in distance:
    if dist != INF:  # INF가 아니라면 카운트
        count += 1
        max_distance = max(max_distance, dist)

print(count - 1, max_distance)  # 자기 자신은 빼야되므로
